lib/active_model/serializer.rb

```diff
@@ -102,24 +102,49 @@ def self.attributes(*attrs)
       end
     end
 
+    # TODO: remove the dynamic method definition
     # @example
     #   class AdminAuthorSerializer < ActiveModel::Serializer
     #     attributes :id, :recent_edits
     #     attribute :name, key: :title
     #
+    #     attribute :full_name do
+    #       "#{object.first_name} #{object.last_name}"
+    #     end
+    #
     #     def recent_edits
     #       object.edits.last(5)
-    #     enr
-    def self.attribute(attr, options = {})
+    #     end
+    def self.attribute(attr, options = {}, &block)
       key = options.fetch(:key, attr)
-      _attributes_keys[attr] = { key: key } if key != attr
-      _attributes << key unless _attributes.include?(key)
+      reader = if block
+                 ->(instance) { instance.instance_eval(&block) }
+               elsif _fragmented
+                 ->(instance) { instance.class._fragmented.read_attribute_for_serialization(attr) }
+               else
+                 ->(instance) { instance.read_attribute_for_serialization(attr) }
+               end
 
-      ActiveModelSerializers.silence_warnings do
-        define_method key do
-          object.read_attribute_for_serialization(attr)
-        end unless method_defined?(key) || _fragmented.respond_to?(attr)
-      end
+      _attribute_mappings[key] = Attribute.new(attr, reader)
+    end
+
+    # @api private
+    # names of attribute methods
+    # @see Serializer::attribute
+    def self._attributes
```

Comment:
I sorta think this and the other `_attribute*` methods should be public, and that the underscore just means that the Serializer owns it.. it could be useful for testing, but keeping it unchanged till we need it public (disclaimer: I've used it for testing)